api/v1/resourceset_types.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 Stefan Prodan.
+// SPDX-License-Identifier: AGPL-3.0
+
+package v1
+
+import (
+	"strings"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/fluxcd/pkg/apis/meta"
+)
+
+const (
+	ResourceSetKind = "ResourceSet"
+)
+
+// ResourceSetSpec defines the desired state of ResourceSet
+type ResourceSetSpec struct {
+	// CommonMetadata specifies the common labels and annotations that are
+	// applied to all resources. Any existing label or annotation will be
+	// overridden if its key matches a common one.
+	// +optional
+	CommonMetadata *CommonMetadata `json:"commonMetadata,omitempty"`
+
+	// Inputs contains the list of resource group inputs.
+	// +optional
+	Inputs []ResourceSetInput `json:"inputs,omitempty"`
+
+	// Resources contains the list of Kubernetes resources to reconcile.
+	// +optional
+	Resources []*apiextensionsv1.JSON `json:"resources,omitempty"`
+
+	// DependsOn specifies the list of Kubernetes resources that must
+	// exist on the cluster before the reconciliation process starts.
+	// +optional
+	DependsOn []Dependency `json:"dependsOn,omitempty"`
+
+	// The name of the Kubernetes service account to impersonate
+	// when reconciling the generated resources.
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// Wait instructs the controller to check the health
+	// of all the reconciled resources.
+	// +optional
+	Wait bool `json:"wait,omitempty"`
+}
+
+// Dependency defines a ResourceSet dependency on a Kubernetes resource.
+type Dependency struct {
+	// APIVersion of the resource to depend on.
+	// +required
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of the resource to depend on.
+	// +required
+	Kind string `json:"kind"`
+
+	// Name of the resource to depend on.
+	// +required
+	Name string `json:"name"`
+
+	// Namespace of the resource to depend on.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Ready checks if the resource Ready status condition is true.
+	// +optional
+	Ready bool `json:"ready,omitempty"`
+
+	// ReadyExpr checks if the resource satisfies the given CEL expression.
+	// +optional
+	ReadyExpr string `json:"readyExpr,omitempty"`
+}
+
+// ResourceSetInput defines the key-value pairs of the resource group input.
+type ResourceSetInput map[string]string
```

Comment:
Why not make it `map[string]any` for allowing nested fields? I think it improves a lot your ability to organize your inputs.